### PR TITLE
Improve datastore sessions

### PR DESCRIPTION
**Pull request by shun-fan**
Added transactions for concurrent session writes.
Cursors for datastore are implemented, adding them back.
Polish up session management using datastore.
Add checkstyle.

### CODE REVIEW

File: bookshelf/google-checks.xml [hunk 1]
```diff
@@ -0,0 +1,193 @@
+<?xml version="1.0"?>
```
Comment (lesv) [inline bookshelf/google-checks.xml]: I'm happy to see this, but shouldn't it go in the main dir and be referenced by all the children?  @shun-fan @tswast

File: bookshelf/pom.xml [hunk 2]
```diff
@@ -95,6 +95,24 @@ Copyright 2015 Google Inc. All Rights Reserved.
       </plugin>
       <plugin>
         <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-checkstyle-plugin</artifactId>
```
Comment (lesv) [inline bookshelf/pom.xml]: This definitely should be in super pom as it only complicates the samples.

File: bookshelf/google-checks.xml (hunk 1 above)
Comment (tswast) [inline bookshelf/google-checks.xml]: google-checks can definitely go in the root and be referenced from the children.

File: bookshelf/pom.xml (hunk 2 above)
Comment (tswast) [inline bookshelf/pom.xml]: When I tried this in java-docs-samples the maven-checkstyle-plugin was not getting run on submodules. Maybe there is a way to do it, but I verified it was not running by introducing style errors intentionally and not seeing an error when running `mvn verify` on the root plugin.

**Comment by lesv** [comment]
LGTM - @shun-fan is looking at converting the login and session handling over to filters.

**Pull request merged**